### PR TITLE
feat: Nested CType

### PR DESCRIPTION
**Pull request by Dudleyneedham**
https://github.com/KILTprotocol/ticket/issues/95 https://github.com/KILTprotocol/ticket/issues/94 https://github.com/KILTprotocol/ticket/issues/96

Added the $ref to the metaschema of `CTypeModel`. 
Had to remove the `Type` of the `CTypeModel` as the `ajv` will take validate accordingly.
Added functionality to validate the nested CType when creating a claim.
Added a new test suite to check when testing the nested CType.

Currently only able to validate nested CTypes with a flat claim. Any more deeply nested claims are not working.  // Done

Maybe out of the scope of the SDK but to spread the nested CTypes properties as keys within the properties. Something to be done on the Demo Client. 

## How to test:

Yarn test
Test Suites: 1 skipped, 19 passed, 19 of 20 total
Tests:       2 skipped, 63 passed, 65 total

## Checklist:

- [x] I have verified that the code works
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: src/ctype/CTypeSchema.ts [hunk 1]
```diff
@@ -105,19 +108,22 @@ export const CTypeModel = {
...
+              type: 'string',
+              format: 'uri-reference',
             },
             format: {
               type: 'string',
-              enum: ['date', 'time', 'uri'],
+              enum: ['date', 'time', 'uri-reference'],
             },
           },
-          required: ['type'],
```
Comment (Dudleyneedham) [inline src/ctype/CTypeSchema.ts]: @tjwelde I had to remove this due to the `ajv` checking the `type`. If it had both a `$ref` and `type` it would use the `type` to validate over the `$ref` 

Maybe, something I could do differently.

File: src/ctype/CTypeSchema.ts [hunk 2]
```diff
@@ -93,6 +93,9 @@ export const CTypeModel = {
       default: 'http://kilt-protocol.org/draft-01/ctype#',
       enum: ['http://kilt-protocol.org/draft-01/ctype#'],
     },
+    title: {
```
Comment (Dudleyneedham) [inline src/ctype/CTypeSchema.ts]: I have added the title. 

This will need to be updated in the Demo Client, so it takes the value `$id` was meant to have.

File: src/ctype/CTypeSchema.ts [hunk 3]
```diff
@@ -105,19 +108,22 @@ export const CTypeModel = {
           properties: {
             type: {
               type: 'string',
-              enum: ['string', 'integer', 'number', 'boolean'],
+              enum: ['string', 'integer', 'number', 'boolean', 'object'],
```
Comment (Dudleyneedham) [inline src/ctype/CTypeSchema.ts]: I had to add the `object` as `$ref` was inside an `object`

File: src/ctype/CTypeUtils.ts [hunk 4]
```diff
@@ -40,12 +40,26 @@ export function verifyClaimStructure(claim: any, schema: any): boolean {
...
+  cType: ICType['schema'],
+  nestedCTypes: Array<ICType['schema']>,
+  claimContents: object
+) {
+  const ajv = new Ajv()
+  ajv.addMetaSchema(CTypeModel)
+  const validated = ajv.addSchema(nestedCTypes).compile(cType)
+  const result = validated(claimContents)
+  return result
 }
```
Comment (Dudleyneedham) [inline src/ctype/CTypeUtils.ts]: Do you think their needs to be a check, similar to `verifySchemaWithErrors` in `CTypeUtils`?

File: src/attestation/Attestation.spec.ts [hunk 5]
```diff
@@ -20,6 +20,7 @@ describe('Attestation', () => {
   const rawCType: ICType['schema'] = {
     $id: 'http://example.com/ctype-1',
     $schema: 'http://kilt-protocol.org/draft-01/ctype#',
+    title: 'test',
```
Comment (Dudleyneedham) [inline src/attestation/Attestation.spec.ts]: I have added `title` to all `CType`s within the test suites.

File: src/claim/Claim.ts [hunk 6]
```diff
@@ -38,6 +38,24 @@ export default class Claim implements IClaim {
...
+  public static fromNestedCTypeClaim(
+    cTypeInput: ICType,
+    nestedCType: Array<ICType['schema']>,
+    claimContents: object,
+    claimOwner: IPublicIdentity['address']
+  ): Claim {
+    if (
+      !CTypeUtils.compileSchema(cTypeInput.schema, nestedCType, claimContents)
+    ) {
+      throw Error('Claim data doesnt match and not valid')
```
Comment (Dudleyneedham) [inline src/claim/Claim.ts]: Do you think the wording is incorrect and doesn't correspond to what is happening if an Error is found?

File: src/ctype/CTypeNested.spec.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    const nestedData = Claim.fromNestedCTypeClaim(
+      nestedCType,
+      [passport.schema, kyc.schema],
+      claimContents,
+      identityAlice.address
+    )
+    // @ts-ignore
+    claimContents.fullName = {}
+
+    // expect(
```
Comment (Dudleyneedham) [inline src/ctype/CTypeNested.spec.ts]: @maudnals could you help with this?

File: src/ctype/CTypeNested.spec.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+          $ref: `${kyc.schema.$id}#/properties/ID`,
+        },
+        number: {
+          $ref: `${kyc.schema.$id}#/properties/number`,
+        },
+        name: {
+          $ref: `${kyc.schema.$id}#/properties/name`,
+        },
+      },
+    }
```
Comment (Dudleyneedham) [inline src/ctype/CTypeNested.spec.ts]: As mentioned within the pull request, I am not able to valid deeply nested `CType`s. Need some help with this, plus I think I need a function that will take the required `properties` of a given `CType` and spread them as shown within the `nested` object.

File: src/ctype/CTypeNested.spec.ts (hunk 7 above)
Comment (Dudleyneedham) [reply src/ctype/CTypeNested.spec.ts]: Done.

File: src/ctype/CTypeUtils.ts [hunk 9]
```diff
@@ -40,12 +40,26 @@ export function verifyClaimStructure(claim: any, schema: any): boolean {
...
+}
+
+export function compileSchema(
+  cType: ICType['schema'],
+  nestedCTypes: Array<ICType['schema']>,
+  claimContents: object
+) {
+  const ajv = new Ajv()
+  ajv.addMetaSchema(CTypeModel)
+  const validated = ajv.addSchema(nestedCTypes).compile(cType)
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/ctype/CTypeUtils.ts]: Agreed.

File: src/attestedclaim/AttestedClaim.spec.ts [hunk 10]
```diff
@@ -19,6 +19,7 @@ function buildAttestedClaim(
   const rawCType: ICType['schema'] = {
     $id: 'http://example.com/ctype-1',
     $schema: 'http://kilt-protocol.org/draft-01/ctype#',
+    title: 'test',
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/attestedclaim/AttestedClaim.spec.ts]: I have put something more meaningful.

File: src/ctype/CTypeUtils.ts [hunk 11]
```diff
@@ -40,12 +40,26 @@ export function verifyClaimStructure(claim: any, schema: any): boolean {
 }
 
 export function getHashForSchema(schema: ICType['schema']): string {
-  return Crypto.hashObjectAsStr(schema)
+  const { $id, ...schemaWithoutID } = schema
```
Comment (weichweich) [inline src/ctype/CTypeUtils.ts]: I think that removing the `$id` could be more explicit. there would also be no `'$id' is assigned a value but never used.` Warning.

some thing like this perhaps?
```typescript
  const { ...schemaClone } = schema
  delete schemaClone.$id
  return Crypto.hashObjectAsStr(schemaClone)
```

File: src/ctype/CTypeUtils.ts [hunk 12]
```diff
@@ -40,12 +40,26 @@ export function verifyClaimStructure(claim: any, schema: any): boolean {
...
+  return Crypto.hashObjectAsStr(schemaWithoutID)
+}
+
+export function compileSchema(
+  cType: ICType['schema'],
+  nestedCTypes: Array<ICType['schema']>,
+  claimContents: object
+) {
```
Comment (weichweich) [inline src/ctype/CTypeUtils.ts]: There is a return type missing here?

Comment (Dudleyneedham) [reply src/ctype/CTypeUtils.ts]: Yes, I have updated this.

File: src/ctype/CTypeUtils.ts (hunk 11 above)
Comment (Dudleyneedham) [reply src/ctype/CTypeUtils.ts]: I have updated, it would make it look cleaner. Maybe it is something we might not want to use.

File: src/ctype/CTypeSchema.ts (hunk 2 above)
Comment (tjwelde) [reply src/ctype/CTypeSchema.ts]: please make a ticket to support these changes in the demo-cient

File: src/ctype/CTypeSchema.ts (hunk 1 above)
Comment (tjwelde) [reply src/ctype/CTypeSchema.ts]: I fixed it with the `oneOf`

File: src/ctype/CTypeUtils.ts (hunk 4 above)
Comment (tjwelde) [reply src/ctype/CTypeUtils.ts]: yes

File: src/ctype/CTypeSchema.ts (hunk 2 above)
Comment (Dudleyneedham) [reply src/ctype/CTypeSchema.ts]: added

File: src/ctype/CType.ts [hunk 13]
```diff
@@ -43,8 +43,19 @@ export default class CType implements ICType {
 
     if (!cTypeInput.hash) {
       this.hash = CTypeUtils.getHashForSchema(this.schema)
+      this.schema.$id = `kilt:ctype:${CTypeUtils.getHashForSchema(this.schema)}`
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/ctype/CType.ts]: I can remove the `0x`, but do you think converting it to `base64` is out of the scope of the ticket. 

I can add a ticket to adjust this.

File: src/ctype/CTypeSchema.ts (hunk 1 above)
Comment (Dudleyneedham) [reply src/ctype/CTypeSchema.ts]: NICE

File: src/ctype/CTypeUtils.ts [hunk 14]
```diff
@@ -40,12 +40,27 @@ export function verifyClaimStructure(claim: any, schema: any): boolean {
 }
 
 export function getHashForSchema(schema: ICType['schema']): string {
-  return Crypto.hashObjectAsStr(schema)
+  const { ...schemaWithoutID } = schema
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/ctype/CTypeUtils.ts]: I took the first option

File: src/ctype/CTypeUtils.ts [hunk 15]
```diff
@@ -40,12 +40,27 @@ export function verifyClaimStructure(claim: any, schema: any): boolean {
...
+  delete schemaWithoutID.$id
+  return Crypto.hashObjectAsStr(schemaWithoutID)
+}
+
+export function compileSchema(
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/ctype/CTypeUtils.ts]: Agreed, changed to `validateNestedSchemas`.

How do you feel about this?

File: src/ctype/CTypeUtils.ts (hunk 4 above)
Comment (Dudleyneedham) [reply src/ctype/CTypeUtils.ts]: Ok, I have updated. 

Let me know if it matches what I would need.

File: src/claim/Claim.ts (hunk 6 above)
Comment (Dudleyneedham) [reply src/claim/Claim.ts]: I have updated, hope this represents it better.

File: src/claim/Claim.ts [hunk 16]
```diff
@@ -38,6 +38,39 @@ export default class Claim implements IClaim {
...
+        cTypeInput.schema,
+        nestedCType,
+        claimContents
+      )
+    ) {
+      throw Error('Nested claim data does not validate against CType')
+    }
+    return new Claim({
+      cTypeHash: cTypeInput.hash,
+      contents: claimContents as IClaim['contents'],
```
Comment (tjwelde) [inline src/claim/Claim.ts:69]: We shouldn't introduce anymore `as` typings

File: src/claim/Claim.ts [hunk 17]
```diff
@@ -38,6 +38,39 @@ export default class Claim implements IClaim {
...
+   * @param claimContents The data inside the [[Claim]].
+   * @param claimOwner The [[PublicIdentity]] of the owner of the [[Claim]].
+   *
+   * @returns A [[Claim]] the owner can use.
+   */
+
+  public static fromNestedCTypeClaim(
+    cTypeInput: ICType,
+    nestedCType: Array<ICType['schema']>,
+    claimContents: object,
```
Comment (tjwelde) [inline src/claim/Claim.ts:55]: `IClaim['contents']` was good. why did you change it to object?

File: src/ctype/CType.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (tjwelde) [reply src/ctype/CType.ts:44]: ok, let's keep the 0x. At least, it is an indicator, that it is a hex string and we can still move to base64 later. Please make a ticket in the ideas for improvement project

[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/ctype/CType.ts:44]: I have added the ticket. 

https://github.com/KILTprotocol/ticket/issues/416

File: src/claim/Claim.ts (hunk 17 above)
Comment (Dudleyneedham) [reply src/claim/Claim.ts:55]: Yeah, I don't know why I did this. 

Thank you.

File: src/claim/Claim.ts (hunk 16 above)
Comment (Dudleyneedham) [reply src/claim/Claim.ts:69]: Sorted

File: src/ctype/CType.ts [hunk 18]
```diff
@@ -57,7 +60,23 @@ export default class CType implements ICType {
     CTypeUtils.errorCheck(cTypeInput)
     this.schema = cTypeInput.schema
     this.owner = cTypeInput.owner
-    this.hash = cTypeInput.hash
+
+    if (!cTypeInput.hash) {
```
Comment (tjwelde) [inline src/ctype/CType.ts:64]: I think this is handled via the `CTypeUtils.errorCheck` and the new `CType.fromSchema` functions

File: src/ctype/CType.ts [hunk 19]
```diff
@@ -58,6 +58,7 @@ export default class CType implements ICType {
     this.schema = cTypeInput.schema
     this.owner = cTypeInput.owner
     this.hash = cTypeInput.hash
+    this.schema.$id = `kilt:ctype:${CTypeUtils.getHashForSchema(this.schema)}`
```
Comment (Dudleyneedham) [inline src/ctype/CType.ts:61]: I added this here. Since we are checking in the error handling. I felt it has already been checked.

**Comment by tjwelde** [comment]
The errorcheck function should check, if the `schema.$id` is correct:
https://github.com/KILTprotocol/sdk-js/pull/222/files#diff-490edeca95e7223eaa1fda2ee51db54fR91

**Comment by tjwelde** [comment]
`schema.$id` should be set, only if it isn't passed to the `fromSchema` method. The constructor shouldn't add the `$id`
